com.avaloq.tools.ddk.xtext.ui.test/src/com/avaloq/tools/ddk/xtext/ui/templates/TemplateProposalProviderHelperTest.xtend

```diff
@@ -0,0 +1,263 @@
+/*******************************************************************************
+ * Copyright (c) 2016 Avaloq Evolution AG and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Avaloq Evolution AG - initial API and implementation
+ *******************************************************************************/
+
+package com.avaloq.tools.ddk.xtext.ui.templates
+
+import com.avaloq.tools.ddk.test.core.AfterAll
+import com.avaloq.tools.ddk.test.core.BeforeAll
+import com.avaloq.tools.ddk.test.core.junit.runners.ClassRunner
+import com.google.inject.Guice
+import org.eclipse.jface.text.IDocument
+import org.eclipse.jface.text.IRegion
+import org.eclipse.jface.text.Position
+import org.eclipse.jface.text.templates.Template
+import org.eclipse.xtext.XtextRuntimeModule
+import org.eclipse.xtext.ui.editor.templates.XtextTemplateContext
+import org.eclipse.xtext.ui.editor.templates.XtextTemplateContextType
+import org.junit.Test
+import org.junit.runner.RunWith
+
+import static org.junit.Assert.assertArrayEquals
+import static org.junit.Assert.assertEquals
+import static org.mockito.Matchers.anyInt
+import static org.mockito.Mockito.mock
+import static org.mockito.Mockito.spy
+import static org.mockito.Mockito.verify
+import static org.mockito.Mockito.when
+
+@RunWith(ClassRunner)
+public class TemplateProposalProviderHelperTest {
+
+  static val SIMPLE_ENUM_VARIABLE_TYPE = new SimpleEnumTemplateVariableResolver().type
+  static val VARIABLE_NAME = "variableName"
+  static val RETURNED_PATTERN = "Pattern returned from createSimpleEnumPattern()"
+
+  static val TEMPLATE_NAME = "TemplateName"
+  static val TEMPLATE_DESCRIPTION = "Template description"
+  static val CONTEXT_TYPE_ID = "context.type.ID"
+  static val IS_AUTO_INSERTABLE = true
+
+  static var IDocument mockDocument
+  static var Position mockPosition
+  static var IRegion mockRegion
+
+  static var XtextTemplateContext templateContext
+
+  static var TemplateProposalProviderHelper helper
+
+  @BeforeAll
+  def void beforeAll() {
+    mockDocument = mock(IDocument)
+    mockPosition = mock(Position)
+    mockRegion = mock(IRegion)
+
+    val templateContextType = new XtextTemplateContextType
+    templateContextType.addResolver(new SimpleEnumTemplateVariableResolver)
+    templateContext = new XtextTemplateContext(templateContextType, mockDocument, mockPosition, null, null)
+
+    helper = Guice.createInjector(new XtextRuntimeModule).getInstance(TemplateProposalProviderHelper)
+
+    when(mockDocument.getLineInformationOfOffset(anyInt)).thenReturn(mockRegion)
+    when(mockDocument.get(anyInt, anyInt)).thenReturn("")
+  }
+
+  @AfterAll
+  def void afterAll() {

```

Comment:
I don't quite understand why this method is needed and why the fields are static. I thought this was because of Before/AfterClass annotations. Now the initialization method is an instance one, so the fields can be this way too, right? Same in other tests

Reply:
The fields are static because @BeforeAll and @AfterAll methods are only called once, but the class is reinstantiated before running each @Test. Hence the @AfterAll method is needed to reset the statics to null and avoid wasting memory.

@BeforeAll and @AfterAll are used in order to set up the test infrastructure once, instead of once per @Test. Is there any reason not to do it this way? (AbstractTest does something similar.)

Reply:
Hm, ok, i think i start to understand. Yes, you can in principle reuse the mocks for all tests if their logic allows. You can keep this method, i guess, since it should not make a big difference. However, pls be aware that setting nulls explicitly is an anti-pattern in java. Trying to affect the garbage collection doesn't work/has negative effect in most of the cases.